src/e84_geoai_common/llm/batch.py

```diff
@@ -0,0 +1,324 @@
+            input_s3_file_url, output_s3_directory_url
+        )
+
+        self._ensure_bucket_exists(
+            input_bucket,
+            create_buckets_if_missing=create_buckets_if_missing,
+        )
+        self._ensure_bucket_exists(
+            output_bucket,
+            create_buckets_if_missing=create_buckets_if_missing,
+        )
+
+        # Upload files if conversations exist
+        if conversations:
+            if inference_cfg is None:
+                inference_cfg = LLMInferenceConfig()
+            model_specific_conversations = self._parse_conversations(conversations, inference_cfg)
+            self._upload_conversations(model_specific_conversations, input_bucket, input_bucket_key)
+
+        request = BatchLLMRequest(
+            roleArn=role_arn,
+            jobName=job_name,
+            modelId=self.llm.model_id,
+            inputDataConfig=InputDataConfig(
+                s3InputDataConfig=S3InputDataConfig(s3Uri=input_s3_file_url)
+            ),
+            outputDataConfig=OutputDataConfig(
+                s3OutputDataConfig=S3OutputDataConfig(s3Uri=output_s3_directory_url)
+            ),
+        )
+
+        # invoke with request
+        response = self.bedrock_client.create_model_invocation_job(
+            **request.model_dump(exclude_none=True)
+        )
+        return response.get("jobArn")
+
+    def _parse_s3_urls(
+        self: Self, input_s3_file_url: str, output_s3_directory_url: str
+    ) -> tuple[str, str, str]:
+        """Return the bucket names and key(s) from an S3 URI."""
+        if not input_s3_file_url.startswith("s3://") or not input_s3_file_url.startswith("s3://"):
+            msg = (
+                f"Invalid S3 URI {input_s3_file_url} or {output_s3_directory_url}."
+                f"Must start with 's3://'."
+            )
+            raise ValueError(msg)
+        if input_s3_file_url.endswith("/"):
+            msg = (
+                f"Invalid Input S3 URI {input_s3_file_url}. Must be a filename,"
+                "but found a directory."
+            )
+            raise ValueError(msg)
+        if not output_s3_directory_url.endswith("/"):
+            msg = (
+                f"Invalid Output S3 URI {output_s3_directory_url}. Must be a directory, but found a"
+                f"filename. Make sure there is a '/' at the end of the url."
+            )
+            raise ValueError(msg)
+        input_parts = input_s3_file_url[5:].split("/", 1)
+        if len(input_parts) <= 1:
+            msg = f"Invalid Input S3 URI {input_s3_file_url}. Must have a file path."
+            raise ValueError(msg)
+        input_bucket = input_parts[0]
+        input_key = input_parts[1]
+
+        output_parts = output_s3_directory_url[5:].split("/", 1)
+        output_bucket = output_parts[0]
+
+        return input_bucket, input_key, output_bucket
+
+    def _ensure_bucket_exists(self: Self, bucket: str, *, create_buckets_if_missing: bool) -> None:
```

Comment:
This seems like a good function to move to the `util` module. If you do that can you make `create_buckets_if_missing` default to `False`?